File: VBAudioRouter/AudioGraphControl/IAudioNodeControl.vb

Imports VBAudioRouter.Controls
Imports Windows.Media.Audio

Namespace AudioGraphControl

    Public Interface IAudioNodeControl
        ReadOnly Property BaseAudioNode As IAudioNode

#Region "State"
        Function Initialize(graph As AudioGraph) As Task
        Sub OnStateChanged(state As GraphState)
#End Region

        Property Canvas As Canvas
    End Interface

    Public Interface IAudioNodeControlInput
        Inherits IAudioNodeControl

        ReadOnly Property IncomingConnector As ConnectorControl
    End Interface

    Public Interface IAudioNodeControlEffect
        Inherits IAudioNodeControl
    End Interface

    Public Interface IAudioNodeControlOutput
        Inherits IAudioNodeControl

        ReadOnly Property OutgoingConnector As ConnectorControl
    End Interface

End Namespace

File: VBAudioRouter/GraphControl/IAudioNodeControl.cs
using Microsoft.UI.Xaml.Controls;
using System.Text;
using VBAudioRouter.Controls;
using Windows.Media.Audio;

namespace VBAudioRouter.GraphControl;

internal interface IAudioNodeControl<out TNode> where TNode : IAudioNode
{
    TNode? GraphNode { get; }
    Canvas? Canvas { get; set; }
}

internal interface IAudioNodeControlFactory<TSelf> where TSelf : IAudioNodeControl<IAudioNode>
{
    static abstract ValueTask<TSelf> CreateAsync(AudioGraph graph);
    static virtual string DisplayName
    {
        get
        {
            var typeName = typeof(TSelf).GetType().Name.AsSpan();

            StringBuilder builder = new();
            for (int i = 0; i < typeName.Length; i++)
            {
                char c = typeName[i];

                if (c == 'N') // ToDo: Better check for "NodeControl"
                    break;

                if (char.IsUpper(c))
                    builder.Append(' ');

                builder.Append(c);
            }

            return builder.ToString();
        }
    }
}

internal interface IAudioInputNodeControl<out TNode> : IAudioNodeControl<TNode>
    where TNode : IAudioInputNode
{
    ConnectorControl OutgoingConnector { get; }
}

internal interface IAudioEffectNodeControl : IAudioNodeControl<AudioSubmixNode>
{
}

internal interface IAudioOutputNodeControl<out TNode> : IAudioNodeControl<TNode>
    where TNode : IAudioNode
{
    ConnectorControl IncomingConnector { get; }
}